internal/util/kubernetes/object/status/queue_test.go

```diff
@@ -133,3 +133,78 @@ var (
 		Kind:    "UDPIngress",
 	}
 )
+
+func TestQueuePublish(t *testing.T) {
+	const testBufferSize = 1
+	testObj := &netv1.Ingress{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Ingress",
+			APIVersion: "networking.k8s.io/v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: corev1.NamespaceDefault,
+			Name:      "ingress-test-1",
+		},
+	}
+
+	// shouldCompleteAlmostImmediately is a helper function that runs a given action
+	// in a goroutine and verifies that it completes within a second.
+	shouldCompleteAlmostImmediately := func(t *testing.T, action func()) {
+		done := make(chan struct{})
+		go func() {
+			action()
+			close(done)
+		}()
+		select {
+		case <-done:
+			return
+		case <-time.After(1 * time.Second):
+			t.Fatal("action did not complete in time")
+		}
+	}
+
+	t.Run("does not block when no subscription exists", func(t *testing.T) {
+		q := NewQueue(WithBufferSize(testBufferSize))
+
+		shouldCompleteAlmostImmediately(t, func() {
+			// Publish more events than the buffer size and expect no block.
+			for i := 0; i < testBufferSize+1; i++ {
+				q.Publish(testObj)
+			}
+		})
+	})
+
+	t.Run("blocks when subscription exists and buffer is full", func(t *testing.T) {
+		q := NewQueue(WithBufferSize(testBufferSize))
+		sub := q.Subscribe(testObj.GroupVersionKind())
+
+		shouldCompleteAlmostImmediately(t, func() {
+			// Publish exactly the number of events that fit in the buffer. Expect no block.
+			// This is to ensure that the buffer is full.
+			for i := 0; i < testBufferSize; i++ {
+				q.Publish(testObj)
+			}
+		})
+
+		var publishedAt time.Time
+		publishDone := make(chan struct{})
+		go func() {
+			q.Publish(testObj)
+			publishedAt = time.Now()
+			close(publishDone)
+		}()
+
+		// Give some time for Publish to block.
+		time.Sleep(10 * time.Millisecond)
+
+		// Consume one event to unblock Publish.
+		consumedEventAt := time.Now()
+		<-sub
```

Comment:
Wouldn't this be equivalent without the sleep

```suggestion
		// Consume one event to unblock Publish.
		<-sub
		consumedEventAt := time.Now()
```

Reply:
Unfortunately not, this would be flaky (`publishedAt` assignment can happen before `consumedEventAt`) - proved it by running the test with `-count=100000`. 

I changed the test logic to not rely on time - instead, I just verify that the publish goroutine doesn't return immediately and that after we consume one event and the goroutine returns, the channel gets full again: eea6f04